src/lib.rs

```diff
@@ -3,6 +3,39 @@ pub mod encode;
 use encode::Color;
 
 
+#[derive(Debug, PartialEq)]
+pub struct Code {
+    cmds: Vec<Cmd>,
+}
+
+impl Code {
+    fn new(code: Vec<Cmd>) -> Result<Code, &'static str> {
+        use self::Cmd::*;
+
+        let mut stack_size = 0 as i32;
+        for cmd in code.iter() {
+            stack_size += match *cmd {
+                Var | NumF(_) | NumI(_) => 1,
+                Fg(_) | Bg(_) | Khz(_) | Comment(_) => continue,
+                // These all pop 1 value off the stack and push 1
+                // value back on, so the net effect is no stack change
+                Sin | Cos | Tan => 0,
+                // Arr(x) pops a value off the stack (called the index)
+                // then pops x more values off the stack. Finally, it
+                // pushes one value back onto the stack based on the index
+                // Thus the net effect of Arr is to reduce the stack size by x.
+                Arr(x) => -(x as i32),
+                Cond => -2,
+                _ => -1,
+            };
+            if stack_size <= 0 {
+                return Err("Invalid bytebeat");
```

Comment:
We should use a more useful error type, we could include the instruction that went wrong, what position it was in, how much stack it needed, and how much stack it had, etc…